README
Botcon - An instant messenger bot with machine remote control
=============================================================

ABOUT:
======

Botcon allows you to execute a shell remotely through a xmmp connection. Using a
common xmpp server account, you can add the bot as your friend and send commands
to him that will be by passed to a shell on the machine that Botcon is running.

Before you can access the shell, you need to make login inside the bot, sending
the follow command:

	#login <key>

Where <key> is the access key (password) provided as a argument of Botcon. For
more information just type:

	botcon --help


HOW TO INSTALL:
===============

	./configure
	make
	make install


NOTES:
======

This is just a proof of concept. Botcon it's really in a alpha version and 
should not be used in a production system, and never as a "ssh" like to make remote
control due security issues.

Botcon it's still unsecure and unstable. Please, use just for tests and for
education purposes.


BUG REPORTS:
============

Please, report to: rene@renesp.com.br



ChangeLog
2011-12-02 Renê S. Pinto <rene@renesp.com.br>

	* botcon:
	First alpha release.



src/xmpp.c
/**
 * BOTCON - An Instant Messenger Bot
 *
 * Copyright (C)2008 Renê de Souza Pinto  (rene at grad.icmc.usp.br)
 *
 *  This file is part of BOTCON, BOTCON is an simple IM bot.
 *
 *  BOTCON is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2.
 *
 *  BOTCON is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 *  License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Bash; see the file COPYING.  If not, write to the Free
 *  Software Foundation, 59 Temple Place, Suite 330, Boston, MA 02111 USA.
 */

#include "xmpp.h"

extern char *proccessMessage(const char *user, const char *msg);

/**
 * Methods
 */
static int  connect(char *server, char *jid, char *passwd);
static void disconnect();
static char *getError();
static int  setState(int state);
static int  getState();
static int  removeUser(char *user);
static int  acceptUser(char *user);
static int  acceptRemove(char *user);
static int  blockUser(char *user);
static int  unblockUser(char *user);
static int  sendMessage(char *user, char *msg);

static void register_handlers();
static void connection_open_cb(LmConnection *con, gboolean result, protocol *info);
static void authentication_cb(LmConnection *con, gboolean result, gpointer user_data);
static LmHandlerResult callback_messages(LmMessageHandler *handler, LmConnection *con, LmMessage *msg, gpointer user_data);
static LmHandlerResult callback_presence(LmMessageHandler *handler, LmConnection *con, LmMessage *msg, gpointer user_data);


/**
 * Private attributes
 */
static protocol *self        = NULL;
static LmConnection *conn    = NULL;
static GError *conn_error    = NULL;

/**
 * instance class XMPP
 */
protocol *new_XMPP()
{
	protocol *newxmpp;

	b_print("new_XMPP(): called\n");

	newxmpp = (protocol*)malloc(sizeof(protocol));
	if(newxmpp == NULL) {
		b_print("new_XMPP(): FAIL: memory allocation failed!\n");
		return(NULL);
	} else {
		newxmpp->state           = ST_UNKNOWN;
		newxmpp->connect         = connect;
		newxmpp->disconnect      = disconnect;
		newxmpp->getError        = getError;
		newxmpp->setState        = setState;
		newxmpp->getState        = getState;
		newxmpp->acceptUser      = acceptUser;
		newxmpp->acceptRemove    = acceptRemove;
		newxmpp->addUser         = NULL;		/* TODO: Implement */
		newxmpp->removeUser      = removeUser;
		newxmpp->blockUser       = blockUser;
		newxmpp->unblockUser     = unblockUser;
		newxmpp->sendMessage     = sendMessage;
		newxmpp->loadUserList    = NULL;			/* TODO: Implement */
		newxmpp->proccessMessage = proccessMessage;
		self = newxmpp;
		b_print("new_XMPP(): OK: object created!\n");
		return(newxmpp);
	}
}


/**
 * Destroy class XMPP
 */
void destroy_XMPP(protocol *obj)
{
	b_print("destroy_XMPP(): called\n");
	free(obj);
	obj = self = NULL;
}


/**
 * Connect on server
 */
static int connect(char *server, char *jid, char *passwd)
{
	b_print("connect(): called\n");

	strcpy(self->server, server);
	strcpy(self->name, jid);
	strcpy(self->passwd, passwd);

	/* Create connection */
	conn = lm_connection_new(server);

	/* Register handlers */
	register_handlers();

	/* Connect */
	if( !lm_connection_open(conn, (LmResultFunction) connection_open_cb, self, NULL, &conn_error) )
		return(FALSE);

	return(TRUE);
}


/**
 * Disconnect from server
 */
static void disconnect()
{
	b_print("disconnect(): called\n");
	lm_connection_close(conn, NULL);
	lm_connection_unref(conn);
	return;
}


/**
 * Get error message from server
 */
static char *getError()
{
	if(conn_error != NULL) {
		return(conn_error->message);
	} else {
		return("unknwon error");
	}
}


/**
 * Set user state
 */
static int  setState(int state)
{
	char *sstate;
	LmMessage *mt;

	b_print("setState(): called\n");

	if(self == NULL)
		return(FALSE);

	switch(state) {
		case ST_UNKNOWN: /* We need to select one, so, select online state */
		case ST_ONLINE:
						sstate = "chat";
						break;

		case ST_AWAY:
						sstate = "away";
						break;

		case ST_BUSY:
						sstate = "dnd";
						break;

		default:
			return(FALSE);
	}

	mt = lm_message_new_with_sub_type(NULL, LM_MESSAGE_TYPE_PRESENCE, LM_MESSAGE_SUB_TYPE_AVAILABLE);
	//lm_message_node_add_child(mt->node, "show", NULL);

	/* Send to server */
	if( !lm_connection_send(conn, mt, &conn_error) ) {
		lm_message_unref(mt);
		b_print("setState(): FAIL: error seting state\n");
		return(FALSE);
	}

	self->state = state;
	lm_message_unref(mt);
	return(TRUE);
}


/**
 * Return state
 */
static int  getState()
{
	if(self == NULL)
		return(FALSE);

	return(self->state);
}


/**
 * Remove user
 */
static int  removeUser(char *user)
{
	LmMessage *mt;

	b_print("removeUser(): called\n");

	mt = lm_message_new_with_sub_type(user, LM_MESSAGE_TYPE_PRESENCE, LM_MESSAGE_SUB_TYPE_UNSUBSCRIBE);

	/* Send to server */
	if( !lm_connection_send(conn, mt, &conn_error) ) {
		lm_message_unref(mt);
		b_print("removeUser(): FAIL: error comunicating with server\n");
		return(FALSE);
	}

	lm_message_unref(mt);
	return(TRUE);
}


/**
 * Accpet remove user
 */
static int  acceptRemove(char *user)
{
	LmMessage *mt;

	b_print("acceptRemove(): called\n");

	mt = lm_message_new_with_sub_type(user, LM_MESSAGE_TYPE_PRESENCE, LM_MESSAGE_SUB_TYPE_UNSUBSCRIBED);

	/* Send to server */
	if( !lm_connection_send(conn, mt, &conn_error) ) {
		lm_message_unref(mt);
		b_print("acceptRemove(): FAIL: error comunicating with server\n");
		return(FALSE);
	}

	lm_message_unref(mt);
	return(TRUE);
}


/**
 * Accpet user invitation
 */
static int  acceptUser(char *user)
{
	LmMessage *mt;

	b_print("acceptUser(): called\n");

	mt = lm_message_new_with_sub_type(user, LM_MESSAGE_TYPE_PRESENCE, LM_MESSAGE_SUB_TYPE_SUBSCRIBED);

	/* Send to server */
	if( !lm_connection_send(conn, mt, &conn_error) ) {
		lm_message_unref(mt);
		b_print("acceptUser(): FAIL: error comunicating with server\n");
		return(FALSE);
	}

	lm_message_unref(mt);
	return(TRUE);
}


/**
 * Block a user
 */
static int  blockUser(char *user)
{
	b_print("blockUser(): called  with %s !!! NOT IMPLEMENTED !!!\n", user);
	return(FALSE);
}


/**
 * Unblock a user
 */
static int  unblockUser(char *user)
{
	b_print("unblockUser(): called  with %s !!! NOT IMPLEMENTED !!!\n", user);
	return(FALSE);
}


/**
 * Send a message to user
 */
static int  sendMessage(char *user, char *msg)
{
	LmMessage *mt;
	GString *str = g_string_new(msg);

	b_print("sendMessage(): called\n");

	mt = lm_message_new_with_sub_type(user, LM_MESSAGE_TYPE_MESSAGE, LM_MESSAGE_SUB_TYPE_CHAT);
	lm_message_node_add_child(mt->node, "body", str->str);

	/* Send to server */
	if( !lm_connection_send(conn, mt, &conn_error) ) {
		lm_message_unref(mt);
		b_print("sendMessage(): FAIL: error comunicating with server\n");
		return(FALSE);
	}

	lm_message_unref(mt);
	g_string_free(str, TRUE);
	return(TRUE);
}


/**
 * Register handlers
 */
static void register_handlers()
{
	LmMessageHandler *handler;

	b_print("register_handlers(): called\n");

	handler = lm_message_handler_new(callback_messages, NULL, NULL);
	lm_connection_register_message_handler(conn, handler, LM_MESSAGE_TYPE_MESSAGE, LM_HANDLER_PRIORITY_NORMAL);
	lm_message_handler_unref(handler);

	handler = lm_message_handler_new(callback_presence, NULL, NULL);
	lm_connection_register_message_handler(conn, handler, LM_MESSAGE_TYPE_PRESENCE, LM_HANDLER_PRIORITY_NORMAL);
	lm_message_handler_unref(handler);
}


/**
 * Callbacks for authentication
 */
static void connection_open_cb(LmConnection *con, gboolean result, protocol *info)
{
	b_print("connection_open_cb(): called\n");
	lm_connection_authenticate(con, info->name, info->passwd, RESOURCE, (LmResultFunction) authentication_cb, info, FALSE, &conn_error);
}
static void authentication_cb(LmConnection *con, gboolean result, gpointer user_data)
{
	b_print("authentication_cb(): called\n");
	if(result == TRUE) {
		self->setState(ST_ONLINE);
	}
}


/**
 * Callback for incoming messages
 */
static LmHandlerResult callback_messages(LmMessageHandler *handler, LmConnection *con, LmMessage *msg, gpointer user_data)
{
	LmMessageNode *no, *body;
	char *user, *strmsg, *ans;
	GString *str = NULL;

	if(lm_message_get_sub_type(msg) == LM_MESSAGE_SUB_TYPE_CHAT) {
		no   = lm_message_get_node(msg);
		body = lm_message_node_get_child(no, "body");
		if(body != NULL) {
			user   = (char*)lm_message_node_get_attribute(no, "from");
			strmsg = (char*)lm_message_node_get_value(body);

			b_print("received message from: %s\n", user);

			/* Proccess message */
			ans = self->proccessMessage(user, strmsg);
			if(ans != NULL) {
				str = g_string_new(ans);
				self->sendMessage(user, str->str);
			}

			lm_message_node_unref(body);
		}
		lm_message_node_unref(no);

		g_string_free(str, TRUE);
		str = NULL;
	} else {
		b_print("unknwon type of message received\n");
		return(LM_HANDLER_RESULT_ALLOW_MORE_HANDLERS);
	}
	return(LM_HANDLER_RESULT_REMOVE_MESSAGE);
}


/**
 * Callback for presence messages
 */
static LmHandlerResult callback_presence(LmMessageHandler *handler, LmConnection *con, LmMessage *msg, gpointer user_data)
{
	LmMessageNode *no;
	LmMessageType type;
	char *user;

	/* Message info */
	type = lm_message_get_sub_type(msg);
	no   = lm_message_get_node(msg);
	if(no != NULL) {
		user = (char*)lm_message_node_get_attribute(no, "from");
	} else {
		user = NULL;
	}

	/* Actions */
	if(type == LM_MESSAGE_SUB_TYPE_SUBSCRIBE) {

		b_print("received invitation from: %s", user);

		// TODO: Check user in blacklist
		self->acceptUser(user); /* For while, just accept any user */

		lm_message_node_unref(no);
		return(LM_HANDLER_RESULT_REMOVE_MESSAGE);

	} else if(type == LM_MESSAGE_SUB_TYPE_UNSUBSCRIBE) {

		// Somebody does not wanna be our friend, Ok...
		self->acceptRemove(user);
	}

	return(LM_HANDLER_RESULT_ALLOW_MORE_HANDLERS);
}



src/xmpp.h
/**
 * BOTCON - An Instant Messenger Bot
 *
 * Copyright (C)2010 Renê de Souza Pinto  (rene at renesp.com.br)
 *
 *  This file is part of BOTCON, BOTCON is an simple IM bot.
 *
 *  BOTCON is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2.
 *
 *  BOTCON is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 *  License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Bash; see the file COPYING.  If not, write to the Free
 *  Software Foundation, 59 Temple Place, Suite 330, Boston, MA 02111 USA.
 */

#ifndef BOTCON_XMPP

	#define BOTCON_XMPP 1

	#include <stdlib.h>
	#include <string.h>
	#include <strings.h>
	#include <loudmouth/loudmouth.h>
	#include "protocol.h"
	#include "botcon.h"

	#define RESOURCE "Botcon"

	protocol *new_XMPP();
	void destroy_XMPP(protocol *obj);

#endif /* BOTCON_XMPP */



src/botcon.h
/**
 * BOTCON - An Instant Messenger Bot
 *
 * Copyright (C)2010 Renê de Souza Pinto  (rene at renesp.com.br)
 *
 *  This file is part of BOTCON, BOTCON is an simple IM bot.
 *
 *  BOTCON is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2.
 *
 *  BOTCON is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 *  License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Bash; see the file COPYING.  If not, write to the Free
 *  Software Foundation, 59 Temple Place, Suite 330, Boston, MA 02111 USA.
 */

#ifndef BOTCON_H

	#define BOTCON_H 1

	#ifdef HAVE_CONFIG_H
		#include <config.h>
	#else
		#error config.h not found. Are you running configure script?
	#endif

	#include <stdio.h>
	#include <strings.h>
	#include <stdarg.h>
	#include <signal.h>
	#include <getopt.h>
	#include <glib.h>
	#include "protocol.h"
	#include "xmpp.h"

	struct _botcon {
		protocol *im;
		char *name;
		char *user;
		char *password;
		char *server;
		char *info;
		char *key;
		GSList *users;    /* Users logged in */
		GSList *running;  /* Users running proccess */
	};

	struct _userproc {
		GString *user;
		GString *filename;
		int in[2];
		int out[2];
		pid_t child;
	};

	typedef struct _botcon botcon_t;
	typedef struct _userproc userproc_t;

	/* Prototypes */
	void b_print(const char *format, ...);

#endif



src/control.c
/**
 * BOTCON - An Instant Messenger Bot
 *
 * Copyright (C)2008 Renê de Souza Pinto  (rene at renesp.com.br)
 *
 *  This file is part of BOTCON, BOTCON is an simple IM bot.
 *
 *  BOTCON is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2.
 *
 *  BOTCON is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 *  License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Bash; see the file COPYING.  If not, write to the Free
 *  Software Foundation, 59 Temple Place, Suite 330, Boston, MA 02111 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <time.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <signal.h>
#include <regex.h>
#include <unistd.h>
#include <glib.h>
#include "botcon.h"

#define BUFFER_LEN 4096
#define DEFAULT_PROCCESS g_string_new("/bin/bash")

static char *ctrl_help(void);
static char *ctrl_login(const char *user, const char *msg);
static char *ctrl_logout(const char *user);
static char *create_user_proc(GString *user, GString *filename);
static gint mfind_user(gconstpointer a, gconstpointer b);
static gboolean muser_islogged(const char *user);
static gpointer proccess_monitor(gpointer user_proc);
void sigchld_handler(int sig);

extern botcon_t bot;

GString *m_ans = NULL;

/**
 * Function to receive and response messages
 */
char *proccessMessage(const char *user, const char *msg)
{
	GSList *plist;
	userproc_t *proc;
	GString *puser, *com;
	ssize_t bytes;

	/* Response message */
	if (m_ans == NULL)
		m_ans = g_string_new("");
	else
		m_ans = g_string_erase(m_ans, 0, -1);

	/* Check if the user already running a proccess */
	for (plist = bot.running; plist != NULL; plist = g_slist_next(plist)) {
		proc = plist->data;
		if (proc != NULL) {
			puser = proc->user;
			if (strcmp(puser->str, user) == 0) {
				/* Write in pipe's proccess */
				com = g_string_new(msg);
				com = g_string_append(com, "\n");
				bytes = strlen(com->str);
				if (write(proc->in[1], com->str, bytes) < bytes) {
					return("Error to write on proccess pipe.\n");
				} else {
					return(NULL);
				}
				g_string_free(com, TRUE);
			}
		}
	}

	/* Check for a command */
	if (msg[0] == '#') {
		/* Parse the command */

		/* HELP */
		if (strcasecmp(&msg[1], "help") == 0) {
			return(ctrl_help());

		/* LOGIN */
		} else if (strncasecmp(&msg[1], "login", 5) == 0) {
			return(ctrl_login(user, msg));

		/* LOGOUT */
		} else if (strncasecmp(&msg[1], "logout", 6) == 0) {
			return(ctrl_logout(user));

		/* INVALID COMMAND */
		} else {
			m_ans = g_string_append(m_ans, "Ops, invalid command :(");
			return(m_ans->str);
		}
	} else {
		m_ans = g_string_append(m_ans, "Please, send #help to see commands available.");
	}

	return(m_ans->str);
}


/**
 * Show help
 */
static char *ctrl_help(void)
{
	char *msg = "\n\
Hi, I am the BOTCON, an instant messenger BOT for machine remote control. The commands avaliable are: \n\n\
    #login <password> - Login at the BOT\n\
    #logout - Logout from BOT\n\
";
	m_ans = g_string_append(m_ans, msg);
	return(m_ans->str);
}

/**
 * Login at Master BOT
 */
static char *ctrl_login(const char *user, const char *msg)
{
	regex_t rexp;
	regmatch_t match;
	int error, len;
	GString *login, *newuser;
	GSList *element;
	struct sigaction newact;

	regcomp(&rexp, " {1,}(\\w+)", REG_EXTENDED | REG_ICASE);

	error = regexec(&rexp, msg, 1, &match, 0);
	if (error == 0) {
		len   = match.rm_eo - match.rm_so;
		login = g_string_sized_new(len);
		g_string_insert_len(login, 0, (msg+match.rm_so+1), len);

		if (strncasecmp(login->str, bot.key, strlen(bot.key)) == 0) {
			/* Log in */
			newuser = g_string_new(user);
			if ((element = g_slist_find_custom(bot.users, newuser, mfind_user)) != NULL) {
				m_ans = g_string_append(m_ans, "You are already logged in. Please, logout first.");
				return(m_ans->str);
			}
			bot.users = g_slist_append(bot.users, newuser);

			m_ans = g_string_append(m_ans, "Login OK. Executing user proccess...\n");

			m_ans = g_string_append(m_ans, create_user_proc(newuser, DEFAULT_PROCCESS));

			memset(&newact, 0, sizeof(newact));
			newact.sa_handler = sigchld_handler;
			sigaction(SIGCHLD, &newact, NULL);
		} else {
			m_ans = g_string_append(m_ans, "Invalid key :(!");
		}
	} else {
		m_ans = g_string_append(m_ans, ":( Command invalid syntax, use:\n    #login <password>\n");
	}
	
	return(m_ans->str);
}

/**
 * Logout at Master BOT
 */
static char *ctrl_logout(const char *user)
{
	GString *olduser;
	GSList *element;

	olduser = g_string_new(user);
	if ((element = g_slist_find_custom(bot.users, olduser, mfind_user)) != NULL) {
		bot.users = g_slist_remove_link(bot.users, element);
		m_ans = g_string_append(m_ans, "Bye Bye :)");
	} else {
		m_ans = g_string_append(m_ans, "You are not logged in! :(");
	}
	
	return(m_ans->str);
}


/**
 * Create an user proccess
 */
static char *create_user_proc(GString *user, GString *filename)
{
	userproc_t *newproc;
	GString *res;
	pid_t child;
	char *args[] = { NULL, NULL };
	int d1, d2;

	res = g_string_new("");
	newproc = (userproc_t*)g_malloc(sizeof(userproc_t));
	if (newproc != NULL) {
		newproc->user = user;
		newproc->filename = filename;
		
		/* Create pipes */
		if (pipe(newproc->in) < 0 || pipe(newproc->out) < 0) {
			res = g_string_append(res, "Error: could not create pipes.\n");
		}

		/* Make fork */
		if ((child = fork()) == 0) {
			/* Child */

			/* Redirect standard input and ouput */
			close(STDIN_FILENO);
			d1 = dup(newproc->in[0]);
			close(newproc->in[0]);
			close(newproc->in[1]);

			close(STDOUT_FILENO);
			d2 = dup(newproc->out[1]);
			close(newproc->out[0]);
			close(newproc->out[1]);
			
			if (d1 < 0 || d2 < 0) {
				g_warning("Could not redirect input/output descriptors.\n");
			}

			args[0] = newproc->filename->str;
			execvp(args[0], args);

			exit(EXIT_FAILURE);
		} else {
			/* Father */
			close(newproc->in[0]);
			close(newproc->out[1]);

			newproc->child = child;

			bot.running = g_slist_append(bot.running, newproc);

			/* Create thread to forward output proccess */
			g_thread_create(proccess_monitor, newproc, TRUE, NULL);
		}
	}

	return(res->str);
}


/**
 * Check if user is logged in
 */
static gboolean muser_islogged(const char *user)
{
	GString *tmp;
	GSList *element;

	tmp = g_string_new(user);
	if ((element = g_slist_find_custom(bot.users, tmp, mfind_user)) != NULL) {
		return(TRUE);
	} else {
		return(FALSE);
	}
}


/**
 * Function to search user in users list
 */
static gint mfind_user(gconstpointer a, gconstpointer b)
{
	GString *astr = (GString*)a;
	GString *bstr = (GString*)b;

	if (astr != NULL && bstr != NULL) {
		if (astr->str != NULL && bstr->str != NULL) {
			return(strcmp(astr->str, bstr->str));
		} else {
			return(-1);
		}
	} else {
		return(-1);
	}
}


/**
 * Thread to read from pipe
 */
static gpointer proccess_monitor(gpointer user_proc)
{
	userproc_t *uproc = (userproc_t*)user_proc;
	ssize_t bytes;
	char buffer[BUFFER_LEN];
	GString *msg;

	while((bytes = read(uproc->out[0], buffer, BUFFER_LEN)) > 0) {
		buffer[bytes] = '\0';
		msg = g_string_new("\n");
		msg = g_string_append(msg, buffer);
		bot.im->sendMessage(uproc->user->str, msg->str);
		g_string_free(msg, TRUE);
	}

	return NULL;
}

/**
 * Signal handling
 */
void sigchld_handler(int sig)
{
	GSList *plist;
	userproc_t *proc;
	GString *puser;
	int status;

	if (sig != SIGCHLD)
		return;

	/* Check children status */
	for (plist = bot.running; plist != NULL; plist = g_slist_next(plist)) {
		proc = plist->data;
		if (proc != NULL) {
			puser = proc->user;

			if (waitpid(proc->child, &status, WNOHANG) >= 0) {
				if (WIFEXITED(status) || WIFSIGNALED(status)) {
					bot.im->sendMessage(proc->user->str, "Proccess finished.\n");
					bot.running = g_slist_remove_link(bot.running, plist);
					ctrl_logout(proc->user->str);
					g_free(proc);
				}
			}
		}
	}
}



src/protocol.h
/**
 * BOTCON - An Instant Messenger Bot
 *
 * Copyright (C)2010 Renê de Souza Pinto  (rene at renesp.com.br)
 *
 *  This file is part of BOTCON, BOTCON is an simple IM bot.
 *
 *  BOTCON is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2.
 *
 *  BOTCON is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 *  License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Bash; see the file COPYING.  If not, write to the Free
 *  Software Foundation, 59 Temple Place, Suite 330, Boston, MA 02111 USA.
 */

#ifndef BOTCON_PROTOCOL_DEF

	#define BOTCON_PROTOCOL_DEF 1

	/**
	 * States
	 */
	#define ST_ONLINE    1
	#define ST_AWAY      2
	#define ST_BUSY      3
	#define ST_UNKNOWN  -1


	/**
	 * This implementation was chosen because we just need one object per
	 * protocol. Thus, it's unnecessary pass the object pointer to each
	 * method function. Each function it's associated with just one 
	 * object (pointed by "self" variable).
	 * 
	 * See xmpp.c
	 */

	typedef struct skel_protocol {
		int    state;
		char   name[1024];
		char   passwd[1024];
		char   server[1024];
		int    (*connect)(char *, char *, char *);
		void   (*disconnect)(void);
		char*  (*getError)(void);
		int    (*setState)(int);
		int    (*getState)(void);
		int    (*acceptUser)(char *);
		int    (*acceptRemove)(char *);
		int    (*addUser)(char *);
		int    (*removeUser)(char *);
		int    (*blockUser)(char *);
		int    (*unblockUser)(char *);
		int    (*sendMessage)(char *, char *);
		int    (*loadUserList)(void);
		char*  (*proccessMessage)(const char *user, const char *msg);
	} protocol;


#endif /* BOTCON_PROTOCOL_DEF */



src/botcon.c
/**
 * BOTCON - An Instant Messenger Bot for Remote Control
 *
 * Copyright (C)2010 Renê de Souza Pinto  (rene at renesp.com.br)
 *
 *  This file is part of BOTCON, BOTCON is an simple IM bot.
 *
 *  BOTCON is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2.
 *
 *  BOTCON is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 *  License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Bash; see the file COPYING.  If not, write to the Free
 *  Software Foundation, 59 Temple Place, Suite 330, Boston, MA 02111 USA.
 */

#include <string.h>
#include <strings.h>
#include "botcon.h"
#include <glib/gthread.h>

void do_quit(int sig);
void show_help(const char *progname);


/* Protocol used for the bot */
protocol *proto;
/* Show debug messages */
static char debug;
/* bot */
botcon_t bot;

/* Main */
int main(int argc, char **argv)
{
	GMainLoop *main_loop;
	struct sigaction newact;
	static struct option longOpts[] = {
		{"help",     no_argument,       NULL, 'h'},
		{"key",      required_argument, NULL, 'k'},
		{"username", required_argument, NULL, 'u'},
		{"server",   required_argument, NULL, 's'},
		{"password", required_argument, NULL, 'p'},
		{"debug",    no_argument,		NULL, 'd'},
		{NULL, no_argument, NULL, 0}
	};
	const char *optstring = "hk:u:s:p:d";
	int opt, optli;
	char optc, *name, *key, *server, *user, *pass;

	/* Initializa vars */
	name = key = server = user = pass = NULL;

	/* Quit callback */
	memset(&newact, 0, sizeof(newact));
	newact.sa_handler = do_quit;
	sigaction(SIGINT, &newact, NULL);

	/* init gthreads */
	g_thread_init(NULL);

	/* Treat command line */
	optc  = 0x00; 
	debug = 0x00;
	opt   = getopt_long(argc, argv, optstring, longOpts, &optli);
	while(opt != -1) {
		switch(opt) {
			case 'k':
				optc |= 0x02;
				key = strdup(optarg);
				break;

			case 'u':
				optc |= 0x04;
				user = strdup(optarg);
				break;
	
			case 's':
				optc |= 0x08;
				server = strdup(optarg);
				break;

			case 'p':
				optc |= 0x10;
				pass = strdup(optarg);
				break;

			case 'd':
				debug = 1;
				break;

			case 'h':
			default:
				optc |= 0x01;
		}
		opt = getopt_long(argc, argv, optstring, longOpts, &optli);
	}

	/* Check arguments */
	if ((optc & 0x1E) == 0x1E)  {
		/* Create the bot */
		
		/* Start XMPP protocol */
		if( (proto = new_XMPP()) == NULL ) {
			exit(EXIT_FAILURE);
		}

		bot.im       = proto;
		bot.name     = name;
		bot.user     = user;
		bot.server   = server;
		bot.password = pass;
		bot.key      = key;
		//bot.info = get_info();
		bot.users   = g_slist_alloc();
		bot.running = g_slist_alloc();

		/* Connect to the server */
		if(proto->connect(server, user, pass) == FALSE) {
			b_print("Connection failed: %s\n", proto->getError());
			return(EXIT_FAILURE);
		}
	} else if (optc != 0x01) {
		show_help(argv[0]);
		return(EXIT_FAILURE);
	} else if (optc == 0x01) {
		show_help(argv[0]);
		return(EXIT_SUCCESS);
	}

	/* Main loop */
	main_loop = g_main_loop_new(NULL, FALSE);
	g_main_loop_run(main_loop);

	free(bot.name);
	free(bot.user);
	free(bot.server);
	free(bot.password);
	free(bot.key);

	g_slist_free(bot.users);
	g_slist_free(bot.running);

	return(EXIT_FAILURE);
}


/**
 * Disconnect from server and exit program
 */
void do_quit(int sig)
{
	proto->disconnect();
	destroy_XMPP(proto);
	exit(EXIT_SUCCESS);
}


/**
 * Print info/debug messages
 */
void b_print(const char *format, ...)
{
	va_list ap;
	if (debug) {
		va_start(ap, format);
		vfprintf(stdout, format, ap);
		va_end(ap);
	}
}


/**
 * Show help
 */
void show_help(const char *progname)
{
	printf("BOTCON - An Instant Messenger Bot for Remote Control\n\n");
	printf("Use: %s -k <key> -u <username> -s <server> -p <password> [-d] [-h]\n\n", progname);
	printf("    -k, --key        Set the access key\n");
	printf("    -u, --username   XMPP server username\n");
	printf("    -s, --server     XMPP server\n");
	printf("    -p, --password   XMPP server password\n");
	printf("    -d, --debug      Show debug messages\n");
	printf("    -h, --help       Show this help\n\n");
}


